### DISCUSSION TITLE
How to update projection and view matrixes by Uniform buffer?

### DISCUSSION THREAD

**Discussion by RikiLee** [vsg-dev/VulkanSceneGraph#1149] [Q&A] [answered]
The camera holds the projection and view matrixes and I found it was updated every frame by push constant. How should I set to update the camera matrixes by using a uniform buffer? Because I want to change some properties between draw calls, which is implemented by pushing constant. So I  want to update camera matrixes by using a uniform buffer

**Comment by robertosfield** [comment, marked as answer]
The VSG does not pass the Camera View matrix to the GPU, it passes a modelview matrix that is accumulated during the RecordTraversal.  This is done to preserve precision in double ssas long as possible before the final matrix has cast down to float and passed to the GPU.  This technique is how the VSG is able to seamlessly cope with whole earth databases without the precision issues.

For performance reasons you want to at least pass the modelview matrix by push constants as it can vary so often between leaves of the scene graph.  Projection matrix is currently passed by push constant alongside the modelview matrix, but this is something I've considered changing this is a projection matrix to a uniform as this only varies once per view per frame, and possibly a view offset for multiview/stereo applications.

Putting the projection matrix into a uniform would free up push constant space, but still the modelview matrix really needs to be passed as push constant.

All the current built-in ShaderSets assume projection and modelview matrices are push constants so that is another issue that would need to be changed.

If you provide all your own shaders then potentially you could pass your own uniforms but the vsg::State is still geared up to pass the project and modelview matrices as push constants so that class would need adapting to do this.

A simpler option would be to just allocated your push constants above the current first 128bytes used by vsg::State to push the project and modelview matrices.

Also consider use per instance array data or BindDescriptorSet to bind uniforms.

One has to asked, just what you are trying to achieve, rather than how you are going about it.  How many draw commands are we talking about?  How many changes of state between them do you need to make?  How much data do you need to change?

**Reply by RikiLee** [reply]
Thanks for your answer!
I have a bunch of lines to draw, and some of them have the same color, for saving memory, I use the color as the instance data, but the line's thickness may change but not as often as the color, so I want to use push constant to change the line thickness.
My shader may look like this:
`#version 450
layout(lines) in;
layout(triangle_strip, max_vertices = 15) out;
layout(push_constant) uniform UniformBufferObject
{
	float thickness;
	uint xViewport;
	uint yViewport;
	uint firstPrimitive;
} ubo;
`
Because the push constant has a very small size of about 256 Bytes, in some cases I have to use a uniform buffer to update projection and view matrixes, maybe this is not a good design

**Comment by robertosfield** [comment]
In the case of line width this is not something you can control with push constants/uniforms you'll need to do it via the graphics pipeline  configuration or enable dynamic state and override the line width using vsg::SetLineWidht/vkSetLineWidth.

https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/commands/SetLineWidth.h
https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkCmdSetLineWidth.html

The vsgdynamicstate example shows you this is done:

https://github.com/vsg-dev/vsgExamples/blob/master/examples/state/vsgdynamicstate/vsgdynamicstate.cpp

**Reply by RikiLee** [reply]
Sorry, I didn't describe it clearly. I use a geometry shader to draw thick lines with caps

**Reply by robertosfield** [reply]
That's an important detail to miss out.

One approach is to pass the data directly as a per instance attribute or pass a "material index" style index as per instance attribute and packed the width etc. data into a array that you access in the vertex/geometry/fragment shaders.

Alternatively you could use a push constant but just use above the 128bytes the the VSG allocates for the project and modelview matrices.
